library/core/src/slice/sort.rs

```diff
@@ -839,6 +848,18 @@ fn partition_at_index_loop<'a, T, F>(
             return;
         }
 
+        if limit == 0 {
+            heapsort(v, is_less);
+            return;
+        }
+
+        // If the last partitioning was imbalanced, try breaking patterns in the slice by shuffling
+        // some elements around. Hopefully we'll choose a better pivot this time.
+        if !was_balanced {
+            break_patterns(v);
```

Comment:
unsure: This looks like it makes it randomized?  Does it need both randomization *and* the heapsort fallback?  Could we just use one or the other?

Reply:
To be honest, I just copied this from the sort implementation. But the fact that sort does need both the pattern breaking and the heapsort fallback makes it seem to me like they're both needed. Alternatively, we could just remove this check, decrement `limit` unconditionally and initialize it to `2 * ilog2(len)`; this is what the "baseline" implementation of introsort/introselect does (or at least the pseudocode on Wikipedia does it this way).

Reply:
Oh, I see, this is exactly what sort is doing

https://github.com/rust-lang/rust/blob/273c6c3913e4d32c3321a3c92fb2ce32c1db9cb8/library/core/src/slice/sort.rs#L746-L758

Makes sense to do the same thing, I guess.